File: 222-count-complete-tree-nodes/count-complete-tree-nodes.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public int CountNodes(TreeNode root) {
        if (root == null) return 0;

        int leftHeight = GetHeight(root, true);
        int rightHeight = GetHeight(root, false);

        // If the left and right heights are the same, it's a perfect binary tree
        if (leftHeight == rightHeight) {
            return (1 << leftHeight) - 1;
        }

        // If not, recursively count the nodes in the left and right subtrees
        return 1 + CountNodes(root.left) + CountNodes(root.right);
    }

    // Helper function to calculate the height of the tree
    private int GetHeight(TreeNode node, bool isLeft) {
        int height = 0;
        while (node != null) {
            height++;
            node = isLeft ? node.left : node.right;
        }
        return height;
    }
}

File: 222-count-complete-tree-nodes/count-complete-tree-nodes.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var countNodes = function(root) {
    if (!root)return 0;
    const calculateDepth=(node)=>{
        let depth=0;
        while (node){depth++; node=node.left;}return depth;
    };
    const leftDepth=calculateDepth(root.left);
    const rightDepth=calculateDepth(root.right);

    if(leftDepth===rightDepth){return(1<<leftDepth)+countNodes(root.right);}
    else{return (1<<rightDepth)+countNodes(root.left);}
};